[FLINK-21932] Add Python SDK showcase (#7)
apache/flink-statefun-playground: python/showcase/README.md

# StateFun Python SDK Showcase

This project is intended for new StateFun users that would like to start implementing their StateFun application functions using Python.
The tutorial is streamlined and split into a few parts which we recommend to go through a specific order, as lay out below.
Each part is demonstrated with some code snippets plus comments to guide you through the SDK fundamentals.

## Prerequisites

- python3
- pip
- docker
- docker-compose

## Building the example

### Using venv

```
python3 -m venv venv
source venv/bin/activate
pip3 install .
```

## Tutorial Sections

> The [__main__.py](showcase/__main__.py) file demonstrates SDK concepts at length, and highly recommend

Reviewer: ```suggestion
The [__main__.py](showcase/__main__.py) file demonstrates SDK concepts at length, and it is highly recommended
```

> to read trough it. The sections below are copied from that file with some of the comments removed.

Reviewer: type: `trough` -> `through`

async def async_ops(context, message):
    profile = context.storage.user_profile
    if not profile:
        user_name = context.address.id

        profile = UserProfile()
        profile.name = user_name
        profile.favorite_ice_cream = await call_favorite_ice_cream_service(user_name)
        profile.favorite_tv_show = await call_faviorite_show_service(user_name)

        context.storage.user_profile = profile

    # reply to our caller with the computed user profile!
    caller: SdkAddress = context.caller
    context.send(message_builder(target_typename=caller.typename,
                                 target_id=caller.id,
                                 value=profile,
                                 value_type=USER_PROFILE_PROTOBUF_TYPE))

async def call_favorite_ice_cream_service(user_name: str) -> str:
    await asyncio.sleep(0.2)
    return "mint chocolate chip"

async def call_faviorite_show_service(user_name: str) -> str:
    await asyncio.sleep(0.5)
    return "The Office"
```

###  Serving

* First, lets define a simple function that computes personalized greetings messages based on the number of times

> That this function was invoked.

Reviewer: ```suggestion
that this function was invoked.
```

For demonstration purposes, this function prints to console the generated greetings
messages.

```
@functions.bind("showcase/serving", specs=[ValueSpec(name="seen", type=IntType)])
async def serving(context, message: Message):
    seen = context.storage.seen or 0
    seen += 1
    context.storage.seen = seen

    greet_request = message.as_type(GREET_JSON_TYPE)
    name = greet_request["name"]

    # in this example, the id part of the currently executing function
    # will be also equal to that name (context.address.id == name).

    print(f"hello {name}! I've seen you {seen} times!", flush=True)
```

* Create a request-reply handler for the registered functions, which understands how to
decode invocation requests dispatched from StateFun cluster, dispatch to the correct function,
and encode side-effects (e.g. storage updates, or invoking other functions)
as responses to be handled by StateFun.

```
statefun_handler = RequestReplyHandler(functions)
```

* Exposing the handler via HTTP
The code below handles the physical HTTP serving.

> In this case we chose aiohttp, although any other HTTP serving framework will do.

Reviewer: ```suggestion
In this case we chose `aiohttp`, although any other HTTP serving framework will do.
```
